DatadogInternal/Sources/Swizzling/MethodSwizzler.swift

```diff
@@ -6,127 +6,160 @@
-        }
+    /// The hierarchy of swizzling per method.
+    private static var swizzlings: [Method: MethodSwizzling] = [:]
 
-        public static func == (lhs: FoundMethod, rhs: FoundMethod) -> Bool {
-            let methodParity = (lhs.method == rhs.method)
-            let classParity = (NSStringFromClass(lhs.klass) == NSStringFromClass(rhs.klass))
-            return methodParity && classParity
-        }
+    /// lock fro synchronizing `swizzlings` mutuations.
+    private static let lock = NSLock()
 
-        public func hash(into hasher: inout Hasher) {
-            let methodName = NSStringFromSelector(method_getName(method))
-            let klassName = NSStringFromClass(klass)
-            let identifier = "\(methodName)|||\(klassName)"
-            hasher.combine(identifier)
-        }
+    /// Synchronization point to access the swizzling nodes.
+    @discardableResult
+    fileprivate static func sync<T>(block: (inout [Method: MethodSwizzling]) -> T) -> T {
+        lock.lock()
+        defer { lock.unlock() }
+        return block(&swizzlings)
     }
+}
 
-    private var implementationCache: [FoundMethod: IMP] = [:]
-    var swizzledMethods: [FoundMethod] {
-        return Array(implementationCache.keys)
+/// Linked list of swizzled implementations.
+///
+/// This object hold the previous (origin) implementation of a
+/// method, the override closure reference, and a reference to its
+/// parent closure.
+private final class MethodSwizzling {
+    /// original implementation
+    let origin: IMP
+    /// type-erased override closure
+    let override: OverrideBox
+    /// parent swizzling
+    let parent: MethodSwizzling?
+
+    init(origin: IMP, override: OverrideBox, parent: MethodSwizzling? = nil) {
+        self.origin = origin
+        self.override = override
+        self.parent = parent
     }
+}
 
-    public static func findMethod(with selector: Selector, in klass: AnyClass) throws -> FoundMethod {
-        /// NOTE: RUMM-452 as we never add/remove methods/classes at runtime,
-        /// search operation doesn't have to wrapped in sync {...} although it's visible in the interface
-        var headKlass: AnyClass? = klass
-        while let someKlass = headKlass {
-            if let foundMethod = findMethod(with: selector, in: someKlass) {
-                return FoundMethod(method: foundMethod, klass: someKlass)
-            }
-            headKlass = class_getSuperclass(headKlass)
-        }
-        throw InternalError(description: "\(NSStringFromSelector(selector)) is not found in \(NSStringFromClass(klass))")
+/// Reference to type-erased override closure.
+private final class OverrideBox {
+    let closure: Any
+    init(_ closure: Any) {
+        self.closure = closure
     }
+}
 
-    public init() { }
+open class MethodSwizzler<Signature, Override> {
+    /// List of swizzling managed by this instance.
+    private var overrides: [(method: Method, `override`: OverrideBox)] = []
```

Comment:
Would it make sense to add method to `OverrideBox`?

Otherwise maybe worth having a struct instead of tuple.

Reply:
The `OverrideBox` is only a reference to a closure, and it is part of the `MethodSwizzling` linked-list so I would prefer not to have the same reference to the `Method` repeated in each element of the list.
Could you elaborate on the struct vs. tuple?  to me they are essentially the same thing, I could do a typealias for readability?

Reply:
Yeah - I guess is readablity only. `typealias` is also fine.